crates/core/src/factories/core_publish.rs

```diff
@@ -0,0 +1,247 @@
+//!
+
+use kitsune2_api::*;
+use message_handler::PublishMessageHandler;
+use std::sync::Arc;
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    task::AbortHandle,
+};
+
+#[cfg(test)]
+mod test;
+
+mod message_handler;
+
+/// CorePublish module name.
+pub const PUBLISH_MOD_NAME: &str = "Publish";
+
+/// CorePublish configuration types.
+mod config {
+    /// Configuration parameters for [CorePublishFactory](super::CorePublishFactory).
+    #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Default)]
+    #[serde(rename_all = "camelCase")]
+    pub struct CorePublishConfig {}
+
+    /// Module-level configuration for CorePublish.
+    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct CorePublishModConfig {
+        /// CorePublish configuration.
+        pub core_publish: CorePublishConfig,
+    }
+}
+
+pub use config::*;
+
+/// A production-ready publish module.
+#[derive(Debug)]
+pub struct CorePublishFactory {}
+
+impl CorePublishFactory {
+    /// Construct a new CorePublishFactory.
+    pub fn create() -> DynPublishFactory {
+        Arc::new(Self {})
+    }
+}
+
+impl PublishFactory for CorePublishFactory {
+    fn default_config(&self, config: &mut Config) -> K2Result<()> {
+        config.set_module_config(&CorePublishModConfig::default())?;
+        Ok(())
+    }
+
+    fn validate_config(&self, _config: &Config) -> K2Result<()> {
+        Ok(())
+    }
+    fn create(
+        &self,
+        builder: Arc<Builder>,
+        space_id: SpaceId,
+        fetch: DynFetch,
+        transport: DynTransport,
+    ) -> BoxFut<'static, K2Result<DynPublish>> {
+        Box::pin(async move {
+            let config: CorePublishModConfig =
+                builder.config.get_module_config()?;
+            let out: DynPublish = Arc::new(CorePublish::new(
+                config.core_publish,
+                space_id,
+                fetch,
+                transport,
+            ));
+            Ok(out)
+        })
+    }
+}
+
+type OutgoingPublishOps = (OpId, Url);
+type IncomingPublishOps = (Vec<OpId>, Url);
+
+#[derive(Debug)]
+struct CorePublish {
+    outgoing_publish_ops_tx: Sender<OutgoingPublishOps>,
+    tasks: Vec<AbortHandle>,
+}
+
+impl CorePublish {
+    fn new(
+        config: CorePublishConfig,
+        space_id: SpaceId,
+        fetch: DynFetch,
+        transport: DynTransport,
+    ) -> Self {
+        Self::spawn_tasks(config, space_id, fetch, transport)
+    }
+}
+
+impl Publish for CorePublish {
+    fn publish_ops(
+        &self,
+        op_ids: Vec<OpId>,
+        target: Url,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Insert requests into publish queue.
+            for op_id in op_ids {
+                if let Err(err) = self
+                    .outgoing_publish_ops_tx
+                    .send((op_id, target.clone()))
```

Comment:
Any reason not to inform the other peer of all the op ids at the same time?

Reply:
Good point, will change that.

Reply:
addressed.